### PR TITLE
Make streaming 3d, 6d & skeleton optional.

### PR DESCRIPTION
**Pull request by EmilRosenquist**
You can now choose which components you want to stream. (Timecode is always streamed)
---
Added the following options to the create live link source menu.

image

### CODE REVIEW

**Comment by helmesjo** [comment]
Add spaces to GUI strings while we are at it.

**Comment by mlveggo** [comment]
Lets call 3d -> markers and 6d -> rigid bodies

**Review by mlveggo — APPROVED** [review]

**Comment by helmesjo** [comment]
Default `3d`/`markers` to off. Or default all except `skeleton` to off?

File: Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Private/QTMConnectLiveLinkSource.cpp [hunk 1]
```diff
@@ -36,13 +36,37 @@ QTMConnectLiveLinkSettings QTMConnectLiveLinkSettings::FromString(const FString&
...
+    }
     return settings;
 }
 
 FString QTMConnectLiveLinkSettings::ToString() const
 {
     FString settingsString = FString::Printf(TEXT("IpAddress=\"%s\""), *IpAddress);
     settingsString.Append(FString::Printf(TEXT("AutoDiscover=\"%d\""), AutoDiscover ? TEXT("true") : TEXT("false")));
+    settingsString.Append(FString::Printf(TEXT("Stream3d=\"%d\""), Stream3d ? TEXT("true") : TEXT("false")));
+    settingsString.Append(FString::Printf(TEXT("Stream6d=\"%d\""), Stream6d ? TEXT("true") : TEXT("false")));
```
Comment (helmesjo) [inline Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Private/QTMConnectLiveLinkSource.cpp:64]: Change GUI labels to `Stream Markers` and `Stream Rigid Bodies`.

File: Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Private/QTMConnectLiveLinkSource.cpp [hunk 2]
```diff
@@ -36,13 +36,37 @@ QTMConnectLiveLinkSettings QTMConnectLiveLinkSettings::FromString(const FString&
...
+    settingsString.Append(FString::Printf(TEXT("StreamSkeleton=\"%d\""), StreamSkeleton ? TEXT("true") : TEXT("false")));
```
Comment (helmesjo) [inline Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Private/QTMConnectLiveLinkSource.cpp:65]: Change GUI label to `Stream Skeleton`.

**Comment by mlveggo** [comment]
6dof and skeleton on, markers can be off. Acting usually involves 6dof props.

File: Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Private/QTMConnectLiveLinkSource.cpp [hunk 3]
```diff
@@ -36,13 +36,37 @@ QTMConnectLiveLinkSettings QTMConnectLiveLinkSettings::FromString(const FString&
...
+        if (stream6d == "false")
+            settings.Stream6d = false;
+    }
+    settings.StreamSkeleton = true;
+    FString streamSkeleton;
+    if (!FParse::Value(*settingsString, TEXT("StreamSkeleton="), streamSkeleton))
+    {
+        if (streamSkeleton == "false")
+            settings.StreamSkeleton = false;
+    }
```
Comment (helmesjo) [inline Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Private/QTMConnectLiveLinkSource.cpp]: Default all but `StreamSkeleton` to `false`.

File: Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Public/QTMConnectLiveLinkBlueprint.h [hunk 4]
```diff
@@ -12,6 +12,6 @@ class QTMCONNECTLIVELINK_API UQTMConnectLiveLinkBlueprint : public UBlueprintFun
 {
     GENERATED_BODY()
 
-    UFUNCTION(BlueprintCallable, Category = "Qualisys", meta = (DisplayName = "Create QTM Connect LiveLink Source", IpAddress = "127.0.0.1", AutoDiscover = "true"))
-    static void CreateQTMConnectLiveLinkSource(FString IpAddress, bool AutoDiscover, FLiveLinkSourceHandle& SourceHandle);
+    UFUNCTION(BlueprintCallable, Category = "Qualisys", meta = (DisplayName = "Create QTM Connect LiveLink Source", IpAddress = "127.0.0.1", AutoDiscover = "true", Stream3d = "true", Stream6d = "true", StreamSkeleton = "true"))
+    static void CreateQTMConnectLiveLinkSource(FString IpAddress, bool AutoDiscover, bool Stream3d, bool Stream6d, bool StreamSkeleton, FLiveLinkSourceHandle& SourceHandle);
```
Comment (helmesjo) [inline Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Public/QTMConnectLiveLinkBlueprint.h]: Default all but `StreamSkeleton` to `false`.

File: Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Public/QTMConnectLiveLinkSource.h [hunk 5]
```diff
@@ -18,7 +18,7 @@ class CRTProtocol;
 class QTMCONNECTLIVELINK_API QTMConnectLiveLinkSettings
 {
 public:
-	QTMConnectLiveLinkSettings() : IpAddress("127.0.0.1"), AutoDiscover(false)
+	QTMConnectLiveLinkSettings() : IpAddress("127.0.0.1"), AutoDiscover(false), Stream3d(true), Stream6d(true), StreamSkeleton(true)
```
Comment (helmesjo) [inline Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Public/QTMConnectLiveLinkSource.h]: Default all but `StreamSkeleton` to `false`.

**Review by helmesjo — CHANGES_REQUESTED** [review]

File: Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Private/QTMConnectLiveLinkSource.cpp (hunk 1 above)
Comment (mlveggo) [reply Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Private/QTMConnectLiveLinkSource.cpp:64]: That is not a gui label. The settings are sent as a string between the livelink editor and livelink plugin. Don't need to change internal naming.

File: Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Private/QTMConnectLiveLinkSource.cpp (hunk 2 above)
Comment (mlveggo) [reply Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Private/QTMConnectLiveLinkSource.cpp:65]: This is not a gui string

File: Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Private/QTMConnectLiveLinkSource.cpp (hunk 3 above)
Comment (mlveggo) [reply Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Private/QTMConnectLiveLinkSource.cpp]: Rather stream3d to false... so we don't default stream markers.

File: Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Public/QTMConnectLiveLinkSource.h (hunk 5 above)
Comment (mlveggo) [reply Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Public/QTMConnectLiveLinkSource.h]: Rather stream3d to false

Comment (helmesjo) [reply Qualisys/QTMConnectLiveLink/Source/QTMConnectLiveLink/Public/QTMConnectLiveLinkSource.h]: Why was this marked as resolved? stream3d is still true. Or is the default value set elsewhere?

**Comment by EmilRosenquist** [comment]
> Add spaces to GUI strings while we are at it.

> Lets call 3d -> markers and 6d -> rigid bodies

Fixed in commits [d025dac3](https://github.com/qualisys/QTM-Connect-For-Unreal/pull/9/commits/d025dac38ce9cfc980715482d97ea123330137db) and [638a295](https://github.com/qualisys/QTM-Connect-For-Unreal/pull/9/commits/638a2958ccbe9e7934a623215607b4a978948611).

> Default `3d`/`markers` to off. Or default all except `skeleton` to off?

> 6dof and skeleton on, markers can be off. Acting usually involves 6dof props.

Fixed in commit [638a295](https://github.com/qualisys/QTM-Connect-For-Unreal/pull/9/commits/638a2958ccbe9e7934a623215607b4a978948611).

**Review by helmesjo — APPROVED** [review]

**Pull request merged**